alert-common/src/main/java/com/synopsys/integration/alert/common/workflow/combiner/AbstractMessageCombiner.java

```diff
@@ -0,0 +1,156 @@
+/**
+ * alert-common
+ *
+ * Copyright (c) 2019 Synopsys, Inc.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.synopsys.integration.alert.common.workflow.combiner;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.springframework.stereotype.Component;
+
+import com.synopsys.integration.alert.common.SetMap;
+import com.synopsys.integration.alert.common.enumeration.ItemOperation;
+import com.synopsys.integration.alert.common.exception.AlertException;
+import com.synopsys.integration.alert.common.exception.AlertRuntimeException;
+import com.synopsys.integration.alert.common.message.model.ComponentItem;
+import com.synopsys.integration.alert.common.message.model.ContentKey;
+import com.synopsys.integration.alert.common.message.model.LinkableItem;
+import com.synopsys.integration.alert.common.message.model.ProviderMessageContent;
+
+@Component
+public class AbstractMessageCombiner implements MessageCombiner {
+    public final List<ProviderMessageContent> combine(List<ProviderMessageContent> messages) {
+        SetMap<ContentKey, ProviderMessageContent> messagesGroupedByKey = SetMap.createLinked();
+        messages.forEach(message -> messagesGroupedByKey.add(message.getContentKey(), message));
+
+        List<ProviderMessageContent> combinedMessages = new ArrayList<>();
+        for (Set<ProviderMessageContent> groupedMessages : messagesGroupedByKey.values()) {
+            combineGroupedMessages(groupedMessages)
+                .ifPresent(combinedMessages::add);
+        }
```

Comment:
Are we going through a bunch of sets we just went through and added to a collection? It looks like we can instead do this originally, right?